pkg/instance/repository.go

```diff
@@ -110,19 +113,73 @@ func (r repository) Delete(id uint) error {
 	return r.db.Unscoped().Delete(&model.Instance{}, id).Error
 }
 
-func (r repository) FindByGroupNames(names []string, presets bool) ([]*model.Instance, error) {
-	var instances []*model.Instance
+const AdministratorGroupName = "administrators"
 
-	err := r.db.
+type GroupWithInstances struct {
+	Name      string
+	Hostname  string
+	Instances []*model.Instance
+}
+
+func (r repository) FindByGroups(groups []*models.Group, presets bool) ([]GroupWithInstances, error) {
+	groupsByName := make(map[string]*models.Group)
+	for _, group := range groups {
+		groupsByName[group.Name] = group
+	}
+	groupNames := maps.Keys(groupsByName)
+
+	instances, err := r.findInstances(groupNames, presets)
+	if err != nil {
+		return nil, err
+	}
+
+	instancesMap := mapInstances(groupNames, instances)
+
+	return groupWithInstances(instancesMap, groupsByName), nil
+}
+
+func (r repository) findInstances(groupNames []string, presets bool) ([]*model.Instance, error) {
+	query := r.db.
 		Preload("RequiredParameters.StackRequiredParameter").
-		Preload("OptionalParameters.StackOptionalParameter").
-		Where("group_name IN ?", names).
+		Preload("OptionalParameters.StackOptionalParameter")
+
+	isAdmin := slices.Contains(groupNames, AdministratorGroupName)
+	if !isAdmin {
+		query = query.Where("group_name IN ?", groupNames)
+	}
+
+	var instances []*model.Instance
+	err := query.
 		Where("preset = ?", presets).
 		Find(&instances).Error
-
 	return instances, err
 }
 
+func mapInstances(groupNames []string, result []*model.Instance) map[string][]*model.Instance {
+	instancesMap := make(map[string][]*model.Instance, len(groupNames))
+	for _, instance := range result {
+		groupName := instance.GroupName
+		instancesMap[groupName] = append(instancesMap[groupName], []*model.Instance{instance}...)
+	}
+	return instancesMap
+}
+
+func groupWithInstances(instancesMap map[string][]*model.Instance, groupMap map[string]*models.Group) []GroupWithInstances {
+	var groupWithInstances []GroupWithInstances
+	for groupName, instances := range instancesMap {
+		if instances == nil {
```

Comment:
Could we get rid of `nil` elements by adding a `where` clause to our query?

Reply:
I don't think so. At least I'm not sure how to do so. We could move the if statement into the "group to instance" function but that would be the same.